File: go/test_performance.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

// ArrayEntry represents an entry in the array-based structure
type ArrayEntry struct {
	Types []string `json:"types"`
	Value int      `json:"value"`
}

// ArrayData represents the array-based structure
type ArrayData struct {
	ObjectClassName string       `json:"objectClassName"`
	LdhName         string       `json:"ldhName"`
	TTL0Data        []ArrayEntry `json:"ttl0_data"`
}

// ObjectData represents the object-based structure
type ObjectData struct {
	ObjectClassName string                    `json:"objectClassName"`
	LdhName         string                    `json:"ldhName"`
	TTL0Data        map[string]map[string]int `json:"ttl0_data"`
}

func getTTLArray(data *ArrayData, recordType string) (int, bool) {
	for _, entry := range data.TTL0Data {
		for _, t := range entry.Types {
			if t == recordType {
				return entry.Value, true
			}
		}
	}
	return 0, false
}

func getTTLObject(data *ObjectData, recordType string) (int, bool) {
	if recordData, ok := data.TTL0Data[recordType]; ok {
		if value, ok := recordData["value"]; ok {
			return value, true
		}
	}
	return 0, false
}

func benchmarkArray(data *ArrayData, recordType string, iterations int) (time.Duration, int) {
	var result int
	start := time.Now()
	for i := 0; i < iterations; i++ {
		result, _ = getTTLArray(data, recordType)
	}
	elapsed := time.Since(start)
	return elapsed, result
}

func benchmarkObject(data *ObjectData, recordType string, iterations int) (time.Duration, int) {
	var result int
	start := time.Now()
	for i := 0; i < iterations; i++ {
		result, _ = getTTLObject(data, recordType)
	}
	elapsed := time.Since(start)
	return elapsed, result
}

func main() {
	iterations := 10000000
	recordType := "A"

	// Load array data
	arrayFile, err := os.ReadFile("/data/array_data.json")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading array_data.json: %v\n", err)
		os.Exit(1)
	}

	var arrayData ArrayData
	if err := json.Unmarshal(arrayFile, &arrayData); err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing array_data.json: %v\n", err)
		os.Exit(1)
	}

	// Load object data
	objectFile, err := os.ReadFile("/data/object_data.json")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading object_data.json: %v\n", err)
		os.Exit(1)
	}

	var objectData ObjectData
	if err := json.Unmarshal(objectFile, &objectData); err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing object_data.json: %v\n", err)
		os.Exit(1)
	}

	// Benchmark array approach
	arrayTime, arrayResult := benchmarkArray(&arrayData, recordType, iterations)

	// Benchmark object approach
	objectTime, objectResult := benchmarkObject(&objectData, recordType, iterations)

	// Convert to milliseconds
	arrayTimeMs := float64(arrayTime.Microseconds()) / 1000.0
	objectTimeMs := float64(objectTime.Microseconds()) / 1000.0

	// Results
	fmt.Println("============================================================")
	fmt.Println("GO PERFORMANCE TEST")
	fmt.Println("============================================================")
	fmt.Printf("Record Type: %s\n", recordType)
	fmt.Printf("Iterations: %d\n", iterations)
	fmt.Printf("\nArray Approach:\n")
	fmt.Printf("  TTL Retrieved: %d\n", arrayResult)
	fmt.Printf("  Total Time: %.2f ms\n", arrayTimeMs)
	fmt.Printf("  Avg Time: %.4f µs/op\n", arrayTimeMs/float64(iterations)*1000)

	fmt.Printf("\nObject Approach:\n")
	fmt.Printf("  TTL Retrieved: %d\n", objectResult)
	fmt.Printf("  Total Time: %.2f ms\n", objectTimeMs)
	fmt.Printf("  Avg Time: %.4f µs/op\n", objectTimeMs/float64(iterations)*1000)

	speedup := arrayTimeMs / objectTimeMs
	fmt.Printf("\nSpeedup: %.2fx (object is %.2fx faster)\n", speedup, speedup)
	fmt.Println("============================================================")
	fmt.Println()
}

File: python/test_performance.py
#!/usr/bin/env python3
import json
import time
import sys

def get_ttl_array(data, record_type):
    """Get TTL from array-based structure"""
    for entry in data['ttl0_data']:
        if record_type in entry['types']:
            return entry['value']
    return None

def get_ttl_object(data, record_type):
    """Get TTL from object-based structure"""
    record_data = data['ttl0_data'].get(record_type)
    if record_data:
        return record_data.get('value')
    return None

def benchmark(data, get_ttl_func, record_type, iterations):
    """Benchmark TTL retrieval"""
    start = time.perf_counter()
    for _ in range(iterations):
        result = get_ttl_func(data, record_type)
    end = time.perf_counter()
    return (end - start) * 1000, result  # Return time in milliseconds

def main():
    iterations = 10000000
    record_type = "A"
    
    # Load array data
    with open('/data/array_data.json', 'r') as f:
        array_data = json.load(f)
    
    # Load object data
    with open('/data/object_data.json', 'r') as f:
        object_data = json.load(f)
    
    # Benchmark array approach
    array_time, array_result = benchmark(array_data, get_ttl_array, record_type, iterations)
    
    # Benchmark object approach
    object_time, object_result = benchmark(object_data, get_ttl_object, record_type, iterations)
    
    # Results
    print("=" * 60)
    print("PYTHON PERFORMANCE TEST")
    print("=" * 60)
    print(f"Record Type: {record_type}")
    print(f"Iterations: {iterations:,}")
    print(f"\nArray Approach:")
    print(f"  TTL Retrieved: {array_result}")
    print(f"  Total Time: {array_time:.2f} ms")
    print(f"  Avg Time: {(array_time/iterations)*1000:.4f} µs/op")
    
    print(f"\nObject Approach:")
    print(f"  TTL Retrieved: {object_result}")
    print(f"  Total Time: {object_time:.2f} ms")
    print(f"  Avg Time: {(object_time/iterations)*1000:.4f} µs/op")
    
    speedup = array_time / object_time
    print(f"\nSpeedup: {speedup:.2f}x (object is {speedup:.2f}x faster)")
    print("=" * 60)
    print()

if __name__ == "__main__":
    main()
